### ISSUE TITLE
possible EventEmitter memory leak detected with mpromise

### ISSUE THREAD

**Issue by whitecolor** [Automattic/mongoose#1992]
```javascript
var def = Model.find({}).exec()

// if we have more then 11 "then waiters"
def.then(function(){
...
})
```

 if we have more then 11 "then waiters" we have an error in console:

```
(node) warning: possible EventEmitter memory leak detected. 11 listeners added. Use emitter.setMaxListeners() to in
crease limit.
Trace
    at Promise.EventEmitter.addListener (events.js:160:15)
    at Promise.on (...\node_modules\mongoose\node_modules\mprom
ise\lib\promise.js:61:31)
    at Promise.onReject (...\node_modules\mongoose\node_modules
\mpromise\lib\promise.js:154:15)
    at Promise.then (...\node_modules\mongoose\node_modules\mpr
omise\lib\promise.js:216:10)
    at Object.<anonymous> (...\app\models\muller-word.js:227:21
)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
```

Not sure if this is expected, but I often use promises in such a way that there may be a lot of listeners for deferred result.

**Comment by danieljoppi** [comment]
+1

**Comment by danieljoppi** [comment]
same problem 

I am use:
 io.js 2.0.0
mongoose 4.0.2

```shell
(node) warning: possible EventEmitter memory leak detected. 11 reconnect listeners added. Use emitter.setMaxListeners() to increase limit.
Trace
    at addListener (events.js:236:17)
    at Server.connect (/Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/lib/server.js:291:17)
    at Db.open (/Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/lib/db.js:190:19)
    at MongoStore._open_database (/Users/daniel.joppi/my-server/node_modules/meanio/lib/core_modules/server/node_modules/connect-mongo/lib/connect-mongo.js:182:15)
    at MongoStore._get_collection (/Users/daniel.joppi/my-server/node_modules/meanio/lib/core_modules/server/node_modules/connect-mongo/lib/connect-mongo.js:177:14)
    at /Users/daniel.joppi/my-server/node_modules/meanio/lib/core_modules/server/node_modules/connect-mongo/lib/connect-mongo.js:194:16
    at /Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/lib/db.js:200:5
    at connectHandler (/Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/lib/server.js:272:7)
    at g (events.js:257:16)
    at emitOne (events.js:77:13)
    at emit (events.js:166:7)
    at /Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:373:23
    at /Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:770:13
    at Callbacks.emit (/Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:84:3)
    at null.messageHandler (/Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:218:23)
    at Socket.<anonymous> (/Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/connection/connection.js:259:22)
(node) warning: possible EventEmitter memory leak detected. 11 destroy listeners added. Use emitter.setMaxListeners() to increase limit.
Trace
    at addListener (events.js:236:17)
    at once (events.js:262:8)
    at connectHandler (/Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/lib/server.js:265:19)
    at g (events.js:257:16)
    at emitOne (events.js:77:13)
    at emit (events.js:166:7)
    at /Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:373:23
    at /Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:770:13
    at Callbacks.emit (/Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:84:3)
    at null.messageHandler (/Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:218:23)
    at Socket.<anonymous> (/Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/connection/connection.js:259:22)
    at emitOne (events.js:77:13)
    at Socket.emit (events.js:166:7)
    at readableAddChunk (_stream_readable.js:145:16)
    at Socket.Readable.push (_stream_readable.js:109:10)
    at TCP.onread (net.js:509:20)
```

**Comment by danieljoppi** [comment]
works with:
 - mongoose 3.8.25
 - io.js 1.8.1

**Comment by mattcasey** [comment]
FYI if you have the connection object, you can setMaxListeners() on it...
```
var connection = mongoose.createConnection(config.MONGO_URI, config.MONGO_OPTIONS);
connection.setMaxListeners(14);
```

**Comment by sham332** [comment]
I am having same issue with Mongoose 4.1.0 or 4.1.1. I don't find this issue with mongoose ~3.8.8 versions.

**Comment by vkarpov15** [comment]
Yeah this issue only happens with mpromise when you do something like this:

```javascript
var p = MyModel.find({}).exec();

for (var i = 0; i < 15; ++i) {
  p.then(function() {
    return new global.Promise(function(resolve) {
      resolve();
    });
  });
}
```

That's unfortunately by design - mpromise uses events internally so every time you call `.then()` that adds an event handler, hence the warning. Fixing mpromise is not a priority going forward since you can set your own promises library in >= 4.1.0. As a workaround in 3.8 you can shut this warning off with `setMaxListeners()` on the promise itself:

```javascript
var p = MyModel.find({}).exec();
p.setMaxListeners(0);

for (var i = 0; i < 15; ++i) {
  p.then(function() {
    return new global.Promise(function(resolve) {
      resolve();
    });
  });
}
```

The recommended solution in 4.1.x though, is to just use your favorite ES6-compatible promises library of choice. mpromise has numerous limitations and issues, and has fundamental flaws that prevent it from working effectively with the `yield` keyword. 

```javascript
// Make mongoose return native promises
mongoose.Promise = global.Promise;

var p = MyModel.find({}).exec();

for (var i = 0; i < 15; ++i) {
  p.then(function() {
    return new global.Promise(function(resolve) {
      resolve();
    });
  });
}
```

**Comment by duke2** [comment]
I try to upgrade mongoose from 3.8.x to 4.x (4.1.10) and i have the same error.

...
danieljoppi stackt trace 
...+ 
C:\Users\igori_000\prj\iioriatti\node_modules\mongoose\node_modules\mongodb\lib\server.js:228
        process.nextTick(function() { throw err; })
                                            ^
Error: connect EADDRINUSE
    at exports._errnoException (util.js:746:11)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1010:19)
[nodemon] app crashed - waiting for file changes before starting...




My project using http://meanjs.org full-stack solution and this problem i have exactly the same when I try to  upgrade one new clean project.

Can you help me to find a way for understand the origin of this problem?
In my code this "p.then ..." is never used and i never have some loop's, when the server start the Mongoose connected to mongodb correctly but when i load the page immediately I have this error.

**Comment by vkarpov15** [comment]
@duke2 that error message has nothing to do with promises.

**Comment by duke2** [comment]
I know because i never use promise directly in my code... but the error is the same and i receve this in one empty mean.js project only when I switch mongose from 3.8.x to 4.x. so I try to ask you some idea's to find the point that generate this problem.

**Comment by vkarpov15** [comment]
Hmm I've never used mean.js so I'm not the right person to ask. But I fail to see how the error you showed is the same as the event emitter memory leak issue mentioned in this thread.

**Comment by duke2** [comment]
I'm sorry... when i write:
...
danieljoppi stackt trace 
...+ 
 i mean that  this replace exactly the stack trace of danieljoppi... 

(node) warning: possible EventEmitter memory leak detected. 11 reconnect listeners added. Use emitter.setMaxListeners() to increase limit.
Trace
    at addListener (events.js:236:17)
    at Server.connect (/Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/lib/server.js:291:17)
    at Db.open (/Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/lib/db.js:190:19)
    at MongoStore._open_database (/Users/daniel.joppi/my-server/node_modules/meanio/lib/core_modules/server/node_modules/connect-mongo/lib/connect-mongo.js:182:15)
    at MongoStore._get_collection (/Users/daniel.joppi/my-server/node_modules/meanio/lib/core_modules/server/node_modules/connect-mongo/lib/connect-mongo.js:177:14)
    at /Users/daniel.joppi/my-server/node_modules/meanio/lib/core_modules/server/node_modules/connect-mongo/lib/connect-mongo.js:194:16
    at /Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/lib/db.js:200:5
    at connectHandler (/Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/lib/server.js:272:7)
    at g (events.js:257:16)
    at emitOne (events.js:77:13)
    at emit (events.js:166:7)
    at /Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:373:23
    at /Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:770:13
    at Callbacks.emit (/Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:84:3)
    at null.messageHandler (/Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:218:23)
    at Socket.<anonymous> (/Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/connection/connection.js:259:22)
(node) warning: possible EventEmitter memory leak detected. 11 destroy listeners added. Use emitter.setMaxListeners() to increase limit.
Trace
    at addListener (events.js:236:17)
    at once (events.js:262:8)
    at connectHandler (/Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/lib/server.js:265:19)
    at g (events.js:257:16)
    at emitOne (events.js:77:13)
    at emit (events.js:166:7)
    at /Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:373:23
    at /Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:770:13
    at Callbacks.emit (/Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:84:3)
    at null.messageHandler (/Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:218:23)
    at Socket.<anonymous> (/Users/daniel.joppi/my-server/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/connection/connection.js:259:22)
    at emitOne (events.js:77:13)
    at Socket.emit (events.js:166:7)
    at readableAddChunk (_stream_readable.js:145:16)
    at Socket.Readable.push (_stream_readable.js:109:10)
    at TCP.onread (net.js:509:20)

For this reason I think that maybe the point in mongoose from where start this type of error you can to know better than me and maybe is possible to improve this point to provide some information that can help to find the origin of this problem.
When I just using mongoose 4.x and I i write some bad code is simple to understand but in the case of one complete migration of one project this type of error can be generate from 1000 reason's i think.
I'm sorry for my English... i hope you can understand.

**Comment by vkarpov15** [comment]
Just use `mongoose.Promise = global.Promise;` (if you're using node 4 and mongoose 4.1) and this issue should go away as specified in https://github.com/Automattic/mongoose/issues/1992#issuecomment-134257407
